src/runtime/uv/tcp.cpp

```diff
@@ -0,0 +1,599 @@
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_listen(b_obj_arg socket, int32_t backlog) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    event_loop_lock(&global_ev);
+
+    int result = uv_listen((uv_stream_t*)tcp_socket->m_uv_tcp, backlog, [](uv_stream_t* stream, int status) {
+        lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket((lean_object*)stream->data);
+
+        if (tcp_socket->m_promise_accept == nullptr) {
+            return;
+        }
+
+        lean_object * promise = tcp_socket->m_promise_accept;
+
+        if (status < 0) {
+            resolve_promise_with_status(promise, status);
+            lean_dec(promise);
+            return;
+        }
+
+        lean_object * client = tcp_socket->m_client;
+        lean_uv_tcp_socket_object * client_socket = lean_to_uv_tcp_socket(client);
+
+        int result = uv_accept((uv_stream_t*)tcp_socket->m_uv_tcp, (uv_stream_t*)client_socket->m_uv_tcp);
+
+        tcp_socket->m_promise_accept = nullptr;
+        tcp_socket->m_client = nullptr;
+
+        if (result < 0) {
+            lean_dec(client);
+            resolve_promise_with_status(promise, result);
+            lean_dec(promise);
+            return;
+        }
+
+        resolve_promise(promise, mk_ok_except(client));
+        lean_dec(promise);
+
+        // The accept increases the count and then the listen decreases
+        lean_dec((lean_object*)stream->data);
+    });
+
+    event_loop_unlock(&global_ev);
+
+    if (result < 0) {
+        return lean_io_result_mk_error(lean_decode_uv_error(result, nullptr));
+    }
+
+    return lean_io_result_mk_ok(lean_box(0));
+}
+
+/* Std.Internal.UV.TCP.Socket.accept (socket : @& Socket) : IO (IO.Promise (Except IO.Error Socket)) */
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_accept(b_obj_arg socket) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    if (tcp_socket->m_promise_accept != nullptr) {
+        return lean_io_result_mk_error(lean_decode_uv_error(UV_EALREADY, mk_string("parallel accept is not allowed! consider binding multiple sockets to the same address and accepting on them instead")));
+    }
+
+    lean_object * promise = create_promise();
+    lean_object * client = unpack_io(lean_uv_tcp_new());
+
+    lean_uv_tcp_socket_object * client_socket = lean_to_uv_tcp_socket(client);
+
+    event_loop_lock(&global_ev);
+    int result = uv_accept((uv_stream_t*)tcp_socket->m_uv_tcp, (uv_stream_t*)client_socket->m_uv_tcp);
+    event_loop_unlock(&global_ev);
+
+    if (result < 0 && result != UV_EAGAIN) {
+        lean_dec(client);
+        resolve_promise_with_status(promise, result);
+    } else if (result >= 0) {
+        resolve_promise(promise, mk_ok_except(client));
+    } else {
+        // The event loop owns the object. It will be released in the listen
+        lean_inc(socket);
+
+        lean_inc(promise);
+        tcp_socket->m_promise_accept = promise;
```

Comment:
We are preparing data for the `uv_listen` callback here while not holding the event loop lock. This means we could potentially have written one of the pointers to non-null already while the other is not valid (in fact it may of course also be that both are not valid data yet). So this can also race.